### ISSUE TITLE
Regression in ActiveRecord 4.2.1 - ActiveRecord::UnknownPrimaryKey when the current table is used as a where clause

### ISSUE THREAD

**Issue by shepmaster** [rails/rails#20602]
## Synopsis

Selecting from a table while using the same table as a `where` clause complains with `ActiveRecord::UnknownPrimaryKey: Unknown primary key for table`:

``` ruby
UserGroup.where(group: UserGroup.all.select(:group_id))
```

This worked correctly in ActiveRecord 4.1.8 and 4.2.0.
## Reproduction

``` ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  # Activate the gem you are reporting the issue against.
  gem 'activerecord', '4.2.1'
  # gem 'activerecord', '4.1.8'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :user_groups, id: false, force: true do |t|
    t.integer :group_id
  end

  create_table :groups, force: true do |t|
    t.integer :post_id
  end
end

class UserGroup < ActiveRecord::Base
  belongs_to :group
end

class Group < ActiveRecord::Base
  has_many :user_groups
end

class BugTest < Minitest::Test
  def test_association_stuff
    group = Group.create!
    user_group = group.user_groups.create!

    groups = UserGroup.where(group: UserGroup.all.select(:group_id))

    assert_equal [user_group.id], groups.map(&:id)
  end
end
```
## Stack trace

```
ActiveRecord::UnknownPrimaryKey: Unknown primary key for table user_groups in model UserGroup.
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/reflection.rb:571:in `primary_key'
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/reflection.rb:315:in `association_primary_key'
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/relation/predicate_builder.rb:68:in `expand'
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/relation/predicate_builder.rb:43:in `block in build_from_hash'
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/relation/predicate_builder.rb:21:in `each'
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/relation/predicate_builder.rb:21:in `build_from_hash'
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/relation/query_methods.rb:960:in `build_where'
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/relation/query_methods.rb:584:in `where!'
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/relation/query_methods.rb:574:in `where'
    /Users/jakegoulding/.rvm/gems/ruby-2.1.5/gems/activerecord-4.2.1/lib/active_record/querying.rb:10:in `where'
    repro.rb:50:in `test_association_stuff'
```

**Comment by shepmaster** [comment]
As a workaround, you can make a more circuitous route and make sure that the expected association class is used:

```ruby
UserGroup.where(group: Group.where(id: UserGroup.all.select(:group_id)))
```

**Comment by shepmaster** [comment]
Note that in the original reproduction, I've disabled the primary key for the join table:

```ruby
create_table :user_groups, id: false, force: true
```

In our case, this is required as we are interfacing with a legacy DB schema we do not control.

If you enable the primary key, you get a different error, this time from the database driver:

```
SQLite3::SQLException: only a single result allowed for a SELECT that is part of an expression
```

The generated SQL includes both the primary key and the selected field:

```sql
SELECT "user_groups".* FROM "user_groups" WHERE "user_groups"."group_id" IN (SELECT "user_groups"."group_id", "user_groups"."id" FROM "user_groups")
```

The same workaround described above helps here.

**Comment by shepmaster** [comment]
An alternate workaround is to use the name of the column, instead of the name of the association:

```ruby
UserGroup.where(group_id: UserGroup.all.select(:group_id))
```

**Comment by timbreitkreutz** [comment]
I was able to reproduce this error for 4.2 and 4.1, but it appears to be working on the master branch...

https://gist.github.com/bcf87feefa838693dfab

**Comment by arthurnn** [comment]
good catch @timbreitkreutz , Can you `git bisect` master and find out which commit fixes the issue?
Maybe it is possible to back port that.

**Comment by timbreitkreutz** [comment]
There was a big refactoring in there, and I was trying to find a back port but after a while I thought since the workaround is so easy, it might not be worth trying to patch the stable 4?  I do have a test case I could submit though for going forward....

**Comment by sgrif** [comment]
If you can just give us the commit ID that fixed it on master we can probably figure out a backport.

**Comment by timbreitkreutz** [comment]
Update:  can no longer run this test case successfully on master. I'm not sure if I made a mistake before or if it regressed again.  Tried git bisect again to locate when the bug was introduced, and ran into many places where Arel appeared to be messed up:

```
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies...
Bundler could not find compatible versions for gem "arel":
  In Gemfile:
    rails (>= 0) ruby depends on
      activerecord (= 5.0.0.alpha) ruby depends on
        arel (~> 6.0) ruby

    arel (>= 0) ruby

```

I'll continue looking in to it.

**Comment by timbreitkreutz** [comment]
I think I might have a fix for this (https://github.com/timbreitkreutz/rails/commit/a37c796e653ea555f81355c9cb732fa1d5e06ed1) -- @sgrif or someone else familiar with this code could you take a look if this makes sense?  Thanks.

Any feedback very welcome--this is my first attempt to contribute to rails.

**Comment by timbreitkreutz** [comment]
If approved I will work on back ports to 4.1 and 4.2.

**Comment by sgrif** [comment]
Please open a pull request and we can discuss there.

**Comment by timbreitkreutz** [comment]
Done... thanks.

**Comment by timbreitkreutz** [comment]
False alarm--I was just reviewing the problem report and notice that I only tested one half.  Need to do the other half now.

**Comment by timbreitkreutz** [comment]
OK I think this one is better @sgrif  et. al.   -- sorry for the hiccup.

**Comment by timbreitkreutz** [comment]
@shepmaster  -- have you had a chance to look at my solution proposal (https://github.com/rails/rails/pull/22251)?  If this solves your situation and you could say so maybe it would help get the fix merged in....  thanks :)

**Comment by shepmaster** [comment]
> had a chance to look at my solution proposal

I skimmed it briefly and it makes sense at a superficial level. Thanks for your work!

> If this solves your situation

Unfortunately, I've moved on from that particular code base and no longer have access to it, sorry!

**Comment by kamipo** [comment]
Related #20802.
Closing since it was fixed in cc0b566.
